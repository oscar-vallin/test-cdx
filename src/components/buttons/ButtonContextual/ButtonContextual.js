import React from 'react';
import PropTypes from 'prop-types';
import { useConst } from '@uifabric/react-hooks';
import { StyledButton } from './ButtonContextual.styles';

const ButtonContextual = ({ id = '__ButtonContext', children, items }) => {
  const menuProps = useConst(() => ({
    shouldFocusOnMount: true,
    items,
  }));

  return (
    <StyledButton menuProps={menuProps} items={items}>
      {children}
    </StyledButton>
  );
};

ButtonContextual.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  items: PropTypes.array,
};

export { ButtonContextual };
