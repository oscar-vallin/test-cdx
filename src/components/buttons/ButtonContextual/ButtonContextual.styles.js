import styled from 'styled-components';
import { DefaultButton } from 'office-ui-fabric-react/lib-commonjs/Button';

export const StyledButton = styled(DefaultButton)`
  background: transparent;
  border: none;
  width: auto;
  padding: 0;
`;
