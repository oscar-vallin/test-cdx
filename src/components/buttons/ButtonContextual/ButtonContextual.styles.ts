import styled from 'styled-components';
import { DefaultButton } from 'office-ui-fabric-react/lib-commonjs/Button';

export const StyledButton = styled(DefaultButton)<StyledButtonProps>`
  background: transparent;
  border: none;
  width: auto;
  padding: 0;
`;

type StyledButtonProps = {
  items: { id?: string; key: string; text: string; onClick: () => void }[];
  menuProps: any;
};
