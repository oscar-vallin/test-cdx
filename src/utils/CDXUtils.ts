import { format } from 'date-fns';

function isDateTimeValid(d?: Date | null): boolean {
  // Check if this date has a valid time value
  /* eslint-disable no-self-compare */
  return !!(d && d.getTime() === d.getTime());
}

function yyyyMMdd(date?: Date | null): string {
  if (date) {
    return format(date, 'yyyy-MM-dd');
  }
  return '';
}

function yyyyMMdda(date?: Date | null): string {
  if (date) {
    return format(date, 'yyyy-MM-dd a');
  }
  return '';
}

function getEnumByValue(enumType, enumValue?: string) {
  const key = Object.keys(enumType).find((x) => enumType[x] === enumValue);
  if (key) {
    return enumType[key];
  }
  return null;
}

export { isDateTimeValid, yyyyMMdd, getEnumByValue, yyyyMMdda };
