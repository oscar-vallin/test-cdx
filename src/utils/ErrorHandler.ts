import { useHistory } from 'react-router-dom';
import { ROUTES } from 'src/data/constants/RouteConstants';
import { useSessionStore } from 'src/store/SessionStore';
import { useNotification } from 'src/hooks/useNotification';
import { useApplicationStore } from 'src/store/ApplicationStore';
import { useLogoutUseCase } from 'src/use-cases/Authentication';

export const ErrorHandler = () => {
  const Toast = useNotification();
  const SessionStore = useSessionStore();
  const ApplicationStore = useApplicationStore();
  const { performUserLogout } = useLogoutUseCase();
  const history = useHistory();

  return (error?: any) => {
    if (error) {
      if (error?.networkError?.statusCode === 403) {
        SessionStore.setGlobalError('Your session has expired please login again.');
        performUserLogout();
      } else {
        const { message } = error;
        const { extensions = null } = error?.graphQLErrors?.shift() || {};

        if (extensions) {
          if (extensions.errorSubType === 'NEED_AUTH') {
            SessionStore.setGlobalError(message);
            performUserLogout();
          } else if (extensions.errorSubType === 'INSUFFICIENT_PRIVILEGES') {
            history.push(ROUTES.ROUTE_UNAUTHORIZED.URL);
          } else {
            Toast.error({ text: 'An internal server error has occurred. Please contact your administrator.' });
          }
        } else if (message === 'Failed to fetch') {
          ApplicationStore.setIsOffline(true);
        } else {
          Toast.error({ text: 'An internal server error has occurred. Please contact your administrator.' });
        }
      }
    }
  };
};
