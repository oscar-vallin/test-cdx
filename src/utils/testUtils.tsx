import { shallow, mount, render } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '../styles/themes/index';
import { theme } from '../styles/themes/theme';

// eslint-disable-next-line react/prop-types
const LightTheme = { ...theme, ...defaultTheme };
export const mountWithTheme = (children: any) => mount(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export const shallowWithTheme = (children: any) =>
  shallow(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export const renderWithTheme = (children: any) => render(<ThemeProvider theme={LightTheme}>{children}</ThemeProvider>);

export * from '@testing-library/react';

export const randomString = (length: number) => {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};
