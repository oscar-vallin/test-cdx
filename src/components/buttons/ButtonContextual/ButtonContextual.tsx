import { ReactElement, ReactNode } from 'react';
import { useConst } from '@uifabric/react-hooks';
import { IContextualMenuProps, IContextualMenuItem } from '@fluentui/react';
import { StyledButton } from './ButtonContextual.styles';

type ButtonContextualProps = {
  id?: string;
  children?: ReactNode;
  title: string;
  items: IContextualMenuItem[];
};

const ButtonContextual = ({ id, children, title, items }: ButtonContextualProps): ReactElement => {
  const menuProps: IContextualMenuProps = useConst(() => ({
    shouldFocusOnMount: true,
    items,
  }));

  return (
    <StyledButton id={id} menuProps={menuProps} title={title} ariaLabel={title} items={items}>
      {children}
    </StyledButton>
  );
};

export { ButtonContextual };
