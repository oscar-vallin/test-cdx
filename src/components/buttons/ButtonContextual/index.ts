import { ButtonContextual } from './ButtonContextual';

export { ButtonContextual };
