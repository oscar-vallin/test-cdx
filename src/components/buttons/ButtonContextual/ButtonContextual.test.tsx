import { mountWithTheme, shallowWithTheme } from 'src/utils/testUtils';
import { ButtonContextual } from './ButtonContextual';
import { IContextualMenuItem } from '@fluentui/react';

const itemsData: IContextualMenuItem[] = [
  {
    id: '1',
    key: 'ProfileMenu_UserSettings',
    text: 'Settings',
    onClick: jest.fn(),
  },
  { id: '2', key: 'ProfileMenu_Logout', text: 'Logout', onClick: jest.fn() },
];

describe('ButtonContextual', () => {
  const tree = shallowWithTheme(
    <ButtonContextual id="__ButtonContextual" title="Profile menu" items={itemsData}>
      Button Contextual
    </ButtonContextual>
  );

  it('Should be defined', () => {
    expect(ButtonContextual).toBeDefined();
  });

  it('Should render correctly', () => {
    expect(tree).toMatchSnapshot();
  });

  it('Should have items', () => {
    expect(tree.children().props().items).toEqual(itemsData);
  });

  it('Should renders children', () => {
    expect(tree.children().contains('Button Contextual')).toEqual(true);
  });

  it('Should render title', () => {
    expect(tree.children().props().title).toEqual('Profile menu');
  });

  it('Should render aria label', () => {
    expect(tree.children().props().ariaLabel).toEqual('Profile menu');
  });

  it('Should renders children when passed in', () => {
    const wrapper = shallowWithTheme(
      <ButtonContextual id="" title="Profile Menu" items={[]}>
        <div className="children" />
      </ButtonContextual>
    );
    expect(wrapper.contains(<div className="children" />)).toEqual(true);
  });

  it('Test styled ButtonContextual component', () => {
    const wrapper = shallowWithTheme(
      <ButtonContextual id="" title="Profile Menu" items={[]}>
        <div className="children" />
      </ButtonContextual>
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('Default props values, Should have a props', () => {
    const defaultTree = shallowWithTheme(<ButtonContextual id={'__ButtonContextual'} title="Profile Menu" items={itemsData} />);
    const mountTree = mountWithTheme(<ButtonContextual id={'__ButtonContextual'} title="Profile Menu" items={itemsData} />);
    expect(defaultTree).toMatchSnapshot();

    expect(defaultTree.children().props().id).toEqual('__ButtonContextual');
    defaultTree.children().simulate('click');
    mountTree.children().simulate('click');
  });
});
