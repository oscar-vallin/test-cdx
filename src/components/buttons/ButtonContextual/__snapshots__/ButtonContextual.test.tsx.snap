// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonContextual Should render correctly 1`] = `ShallowWrapper {}`;
